src/validateAtoms.ts

```diff
@@ -0,0 +1,89 @@
+import { atom, Getter, SetStateAction, WritableAtom } from 'jotai';
+import { loadable } from 'jotai/utils';
+
+import type { CommonState } from './atomWithValidate';
+
+type inferGeneric<Type> = Type extends AtomWithValidation<infer X> ? X : never;
+
+export type Validator<Keys extends symbol | string | number, Vals> = (
+  values: Record<Keys, Vals>,
+) => void | Promise<void>;
+
+export type ValidatorState = {
+  isValid: undefined | boolean;
+  error: null | Error | unknown;
+  isValidating: undefined | boolean;
+};
+
+type AtomWithValidation<Value> = WritableAtom<
+  CommonState<Value>,
+  SetStateAction<Value>
+>;
+
+type State<Values extends Record<string, unknown>> = {
+  values: Values;
+} & ValidatorState;
+
+export const validateAtoms = <
+  AtomGroup extends Record<string, AtomWithValidation<any>>,
+  Keys extends keyof AtomGroup,
+  Vals extends AtomGroup[Keys],
```

Comment:
Could this be the other way around?
```suggestion
  Vals extends Record<string, unknown>,
  AtomGroup extends { [key in keyof Vals]: AtomWithValidation<Vals[key]>>,
```

and in the code:
```ts
type Keys = keyof Vals;
```

Reply:
I think I tried this and TS wasn't inferring the passed object's types, I'll try it again with the above suggestion and ping back

Reply:
I think it can actually be something like this:
```ts
const validateAtoms = <Vals>(
  labeledAtoms: { [key in keyof Vals]: AtomWithValidation<Vals[key]>>,
  validator: Validator<Vals>,
) => {
```
hm, maybe you are right, it can't infer types from `labeledAtoms`.

Reply:
Yeah,
Both won't work, the type inheritance needs it to point on the passed object to infer it as a generic

Reply:
makes sense.